akka-remote/src/main/scala/akka/remote/artery/tcp/SSLEngineProvider.scala

```diff
@@ -84,19 +85,37 @@ class ConfigSSLEngineProvider(protected val config: Config, protected val log: M
   val SSLRequireMutualAuthentication: Boolean = config.getBoolean("require-mutual-authentication")
   val HostnameVerification: Boolean = config.getBoolean("hostname-verification")
 
-  private lazy val sslContext: SSLContext = {
-    // log hostname verification warning once
-    if (HostnameVerification)
-      log.debug("TLS/SSL hostname verification is enabled.")
-    else
-      log.info(
-        LogMarker.Security,
-        "TLS/SSL hostname verification is disabled. See Akka reference documentation for more information.")
-
-    constructContext()
+  private val SSLContextCacheTime: FiniteDuration = config.getDuration("ssl-context-cache-ttl").toMillis.millis
+
+  // log hostname verification warning once
+  if (HostnameVerification)
+    log.debug("TLS/SSL hostname verification is enabled.")
+  else
+    log.info(
+      LogMarker.Security,
+      "TLS/SSL hostname verification is disabled. See Akka reference documentation for more information.")
+
+  private def sslContext: SSLContext = getFromCache
```

Comment:
This is called when a new connection is created. Do we also intend to reconnect existing connections (that can be very long-lived) when the keys change? Maybe TLS will do that automatically?

Reply:
Hmmm, good call.

In order to prevent an attacker to stay connected too long we need a mechanism to shut the attacker connections once their stolen keys are no longer valid.

I'm not sure who triggers a re-handshake: whether particular versions of the protocol or if it's left to the direct user of the [`SSLEngine`](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLEngine.html#beginHandshake--) but a re-handshake between to peers on a long-standing connection could eventually cause a connection failure if/when the certificates have expired. That failure would be handled somewhere causing the reconnection. This, of course, will only happen if re-handshaking is triggered, and the certificates use reasonable expiration dates.

So ending connections of attackers is a combination of several pieces:

 - when the current certificate will expire
 - when the new certificate will be available
 - when this code will re-read the credentials
 - when the ongoing session in the `SSLEngine` will trigger a handshake causing the session to fail and terminate 
 - handling the reconnection (which should use the latest keys and certificates on both ends)

Reply:
For the record. We will not do anything about existing connections.